engine/src/flutter/impeller/entity/contents/text_shadow_cache.h

```diff
@@ -0,0 +1,113 @@
+/// and holds them for at least one frame.
+///
+/// Additionally, there is an optimization for a single glyph (generally an
+/// Icon) that uses the content itself as a key.
+///
+/// If there was a cheaper method of text frame identity, or a per-glyph caching
+/// system this could be more efficient. As it exists, this mostly ameliorate
+/// severe performance degradation for glyph shadows but does not provide
+/// substantially better performance than Skia.
+class TextShadowCache {
+ public:
+  TextShadowCache() = default;
+
+  ~TextShadowCache() = default;
+
+  /// @brief A key to look up cached glyph textures.
+  struct TextShadowCacheKey {
+    Scalar max_basis;
+    int64_t identifier;
+    bool is_single_glyph;
+    Font font;
+    Sigma sigma;
+
+    TextShadowCacheKey(Scalar p_max_basis,
+                       int64_t p_identifier,
+                       bool p_is_single_glyph,
+                       const Font& p_font,
+                       Sigma p_sigma)
+        : max_basis(p_max_basis),
+          identifier(p_identifier),
+          is_single_glyph(p_is_single_glyph),
+          font(p_font),
+          sigma(p_sigma) {}
+
+    struct Hash {
+      std::size_t operator()(const TextShadowCacheKey& key) const {
+        return fml::HashCombine(key.max_basis, key.identifier,
+                                key.is_single_glyph, key.font.GetHash(),
+                                key.sigma.sigma);
+      }
+    };
+
+    struct Equal {
+      constexpr bool operator()(const TextShadowCacheKey& lhs,
+                                const TextShadowCacheKey& rhs) const {
+        return lhs.max_basis == rhs.max_basis &&
+               lhs.identifier == rhs.identifier &&
+               lhs.is_single_glyph == rhs.is_single_glyph &&
+               lhs.font.IsEqual(rhs.font) && lhs.sigma.sigma == rhs.sigma.sigma;
+      }
+    };
+  };
+
+  /// @brief Mark all glyph textures as unused this frame.
+  void MarkFrameStart();
+
+  /// @brief Remove all glyph textures that were not referenced at least once.
+  void MarkFrameEnd();
+
+  /// @brief Lookup the entity in the cache with the given filter/text contents,
+  ///        returning the new entity to render.
+  ///
+  /// If the entity is not present, render and place in the cache.
+  std::optional<Entity> Lookup(const ContentContext& renderer,
+                               const Entity& entity,
+                               const std::shared_ptr<FilterContents>& contents,
+                               const TextShadowCacheKey&);
+
+  // Visible for testing.
+  size_t GetCacheSizeForTesting() const { return entries_.size(); }
+
+ private:
+  TextShadowCache(const TextShadowCache&) = delete;
+
+  TextShadowCache& operator=(const TextShadowCache&) = delete;
+
+  struct TextShadowCacheData {
+    Entity entity;
+    bool used_this_frame = true;
+    Matrix key_matrix;
+  };
+
+  std::unordered_map<TextShadowCacheKey,
```

Comment:
absl::flat_hash_map is recommended ( https://github.com/abseil/abseil-cpp/blob/bcf4bf315ab25075b51ad1bca9da351aa4579c00/absl/container/flat_hash_map.h#L27 )

Reply:
I can switch this but unfortuately flat_hash_map (https://abseil.io/docs/cpp/guides/container)

> Guarantees an O(1) erase method by returning void instead of an iterator.

I think erasing invalidates the iterator so I can't do the optimization you asked for above. Which would you prefer.

Reply:
If you use `absl::flat_hash_map`, you can use `absl::remove_if` to do this really easily.

Reply:
Or `absl::erase_if` maybe in this case.

Reply:
Done